### PR TITLE
Add custom metadata validation based on JSON schema

### PR DESCRIPTION
**Pull request by yufeih**
Implement custom schema validation using standard JSON schema. The following features are ported from schema validation console:

- `type` --> `type`
- `choice_set` --> `enum`
- `is_multivalue` --> `type == Array`

This JSON schema validation could also be shared with SDP.

Added config `metadataSchema` to point to an external JSON schema file that defines metadata schema.

Two error codes are ported from _schema validation console_: `undefined value` and `unexpected-type`

### CODE REVIEW

File: src/docfx/lib/schema/JsonSchemaValidation.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                        {
+                            ValidateCore(schema.Items, item, errors);
+                        }
+                    }
+                    break;
+
+                case JObject map:
+                    foreach (var (key, value) in map)
+                    {
+                        if (schema.Properties.TryGetValue(key, out var propertySchema))
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchemaValidation.cs]: How to handle `AdditionalProperties` is false or object?
https://json-schema.org/understanding-json-schema/reference/object.html#properties

File: src/docfx/lib/schema/JsonSchema.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class JsonSchema
+    {
+        public JsonSchemaType Type { get; set; }
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchema.cs]: Required properties?
https://json-schema.org/understanding-json-schema/reference/object.html#required-properties

File: src/docfx/lib/schema/JsonSchemaValidation.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/schema/JsonSchemaValidation.cs]: Defer that to SDP work

Comment (live1206) [reply src/docfx/lib/schema/JsonSchemaValidation.cs]: Support it during SDP implementation

File: src/docfx/lib/schema/JsonSchema.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/lib/schema/JsonSchema.cs]: Support it during SDP implementation

File: src/docfx/config/Config.cs [hunk 3]
```diff
@@ -150,6 +150,11 @@ internal sealed class Config
         /// </summary>
         public readonly SourceInfo<string> MonikerDefinition = new SourceInfo<string>(string.Empty);
 
+        /// <summary>
+        /// Get the metadata JSON schema.
```
Comment (live1206) [inline src/docfx/config/Config.cs]:
```suggestion
        /// Get the metadata JSON schema file path.
```

File: src/docfx/build/metadata/MetadataProvider.cs [hunk 4]
```diff
@@ -53,7 +53,10 @@ public MetadataProvider(Config config)
             }
 
             var (errors, metadata) = JsonUtility.ToObject<T>(result);
+
             errors.AddRange(MetadataValidator.Validate(result));
+            errors.AddRange(JsonSchemaValidation.Validate(file.Docset.MetadataSchema, result));
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/metadata/MetadataProvider.cs]: yes, I've merged MetadataValidator into MetadataProvider in subsequent [PR](https://github.com/dotnet/docfx/pull/4429)

File: test/docfx.Test/lib/JsonSchemaTest.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.Docs.Build
+{
+    public class JsonSchemaTest
+    {
+        [Theory]
+
+        // type validation
```
Comment (OsmondJiang) [inline test/docfx.Test/lib/JsonSchemaTest.cs]: we want to make sure all our test cases are consistent with Json .Net behaviors, can we copy some tests from there?

Comment (yufeih) [reply test/docfx.Test/lib/JsonSchemaTest.cs]: sure in another PR

Comment (OsmondJiang) [reply test/docfx.Test/lib/JsonSchemaTest.cs]: Seems Json team is also working on this feature, with a total different way, we may want to hold this PR until we all agree to do that in this way.

cc @LowEntropyBody

**Comment by herohua** [comment]
FYI @kimizhu

**Pull request merged**